Compilation error with strong typed enums in map in combination with namespaces

- What is the issue you have?
Compilation error.

- Please describe the steps to reproduce the issue. Can you provide a small but working code example?
Compiling the small program below results in C2166: l-value specifies const object (see full compiler output below). When putting the map and the to/from_json in the same namespace it compiles. It boilds down to an assignment/construction of the map's value type.
```
#include <json/json.hpp>
#include <map>
#include <string>

namespace bla
{
enum class foo : uint16_t
{
	bar, baz
};
} // comment this out and the lines above main in and it will compile
using foo_map = std::map<foo, std::string>;

void to_json(nlohmann::json& j, const foo_map& m)
{

}

void from_json(const nlohmann::json& j, foo_map& m)
{

}
//}
//using bla::foo_map;
int main(int, char**)
{
	nlohmann::json j;
	foo_map m = j;
}
```

- What is the expected behavior?
It should compile without errors 

- And what is the actual behavior instead?
Compilation errors
```
compiling main.cpp
C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.13.26128/bin/HostX64/x64/cl.exe /nologo /c /EHsc /Od /Zi /MDd "/IC:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\ipp\include" /IC:\Qt\Qt5.9.5\5.9.5\msvc2017_64\include /IC:\Qt\Qt5.9.5\5.9.5\msvc2017_64\include\QtCore /IC:\Qt\Qt5.9.5\5.9.5\msvc2017_64\mkspecs\win32-msvc /IC:\Projekte\monalysis\Entwicklung\Quellen\Projekte\build-scratchpad-VC17ohneQt-Debug\qtc_VC17_91031919-debug\scratchpad.a434f584\qt.headers /IC:\Projekte\thirdparty\boost /IC:\Projekte\thirdparty\fmt /IC:\Projekte\thirdparty\json /IC:\Projekte\thirdparty\nonius\nonius /IC:\Projekte\thirdparty\range-v3\range-v3-vs2015\include /DUNICODE /D_UNICODE /DWIN32 /DQT_CORE_LIB /DBOOST_LIB_DIAGNOSTIC /DBOOST_AUTO_LINK_TAGGED /DBOOST_RESULT_OF_USE_DECLTYPE /DBOOST_QT_PUSHBACK_WORKAROUND "/DBOOST_QT_LISTOF_WORKAROUND=1" "/DWINVER=0x0502" "/D_WIN32_WINNT=0x0502" "/D_WIN32_WINDOWS=0x0502" /FoC:\Projekte\monalysis\Entwicklung\Quellen\Projekte\build-scratchpad-VC17ohneQt-Debug\qtc_VC17_91031919-debug\scratchpad.a434f584\3a52ce780950d4d9\main.cpp.obj C:\Projekte\monalysis\Entwicklung\Quellen\Projekte\scratchpad\main.cpp /TP /FS /permissive- /std:c++latest /Zm200
:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.13.26128\include\utility(268): error C2166: l-value specifies const object
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.13.26128\include\utility(267): note: while compiling class template member function 'std::pair<const _Kty,_Ty> &std::pair<const _Kty,_Ty>::operator =(std::pair<const _Kty,_Ty> &&) noexcept(false)'
        with
        [
            _Kty=bla::foo,
            _Ty=std::string
        ]
C:\Projekte\monalysis\Entwicklung\Quellen\Projekte\monalysis-porsche\thirdparty\json\json/json.hpp(1172): note: see reference to function template instantiation 'std::pair<const _Kty,_Ty> &std::pair<const _Kty,_Ty>::operator =(std::pair<const _Kty,_Ty> &&) noexcept(false)' being compiled
        with
        [
            _Kty=bla::foo,
            _Ty=std::string
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.13.26128\include\xtree(443): note: see reference to class template instantiation 'std::pair<const _Kty,_Ty>' being compiled
        with
        [
            _Kty=bla::foo,
            _Ty=std::string
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.13.26128\include\xtree(739): note: see reference to class template instantiation 'std::_Tree_node<std::pair<const _Kty,_Ty>,std::_Default_allocator_traits<_Alloc>::void_pointer>' being compiled
        with
        [
            _Kty=bla::foo,
            _Ty=std::string,
            _Alloc=std::allocator<std::pair<const bla::foo,std::string>>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.13.26128\include\xtree(738): note: while compiling class template member function 'void std::_Tree_comp_alloc<_Traits>::_Freeheadnode(std::_Tree_node<std::pair<const _Kty,_Ty>,std::_Default_allocator_traits<_Alloc>::void_pointer> *)'
        with
        [
            _Traits=std::_Tmap_traits<bla::foo,std::string,std::less<bla::foo>,std::allocator<std::pair<const bla::foo,std::string>>,false>,
            _Kty=bla::foo,
            _Ty=std::string,
            _Alloc=std::allocator<std::pair<const bla::foo,std::string>>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.13.26128\include\xtree(631): note: see reference to function template instantiation 'void std::_Tree_comp_alloc<_Traits>::_Freeheadnode(std::_Tree_node<std::pair<const _Kty,_Ty>,std::_Default_allocator_traits<_Alloc>::void_pointer> *)' being compiled
        with
        [
            _Traits=std::_Tmap_traits<bla::foo,std::string,std::less<bla::foo>,std::allocator<std::pair<const bla::foo,std::string>>,false>,
            _Kty=bla::foo,
            _Ty=std::string,
            _Alloc=std::allocator<std::pair<const bla::foo,std::string>>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.13.26128\include\xtree(842): note: see reference to class template instantiation 'std::_Tree_comp_alloc<_Traits>' being compiled
        with
        [
            _Traits=std::_Tmap_traits<bla::foo,std::string,std::less<bla::foo>,std::allocator<std::pair<const bla::foo,std::string>>,false>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.13.26128\include\map(81): note: see reference to class template instantiation 'std::_Tree<std::_Tmap_traits<_Kty,_Ty,_Pr,_Alloc,false>>' being compiled
        with
        [
            _Kty=bla::foo,
            _Ty=std::string,
            _Pr=std::less<bla::foo>,
            _Alloc=std::allocator<std::pair<const bla::foo,std::string>>
        ]
C:\Projekte\monalysis\Entwicklung\Quellen\Projekte\scratchpad\main.cpp(212): note: see reference to class template instantiation 'std::map<bla::foo,std::string,std::less<_Kty>,std::allocator<std::pair<const _Kty,_Ty>>>' being compiled
        with
        [
            _Kty=bla::foo,
            _Ty=std::string
        ]
```


- Which compiler and operating system are you using? Is it a [supported compiler](https://github.com/nlohmann/json#supported-compilers)?
MSVC 2017 (15.6.6); MSVC 19.0.24215.1 (CMake output)
Not directly supported.

- Did you use a released version of the library or the version from the `develop` branch?
Release 3.1.2

- If you experience a compilation error: can you [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests)?
Tried but ctest gives error `Test not available without configuration.  (Missing "-C <config>"?)` for every test case. If told how to fix this I'll happily execute the tests.

--- a/include/nlohmann/detail/conversions/from_json.hpp
+++ b/include/nlohmann/detail/conversions/from_json.hpp
@@ -5,6 +5,7 @@
 #include <ciso646> // and, not
 #include <forward_list> // forward_list
 #include <iterator> // inserter, front_inserter, end
+#include <map> // map
 #include <string> // string
 #include <tuple> // tuple, make_tuple
 #include <type_traits> // is_arithmetic, is_same, is_enum, underlying_type, is_convertible
@@ -275,6 +276,25 @@ template<typename BasicJsonType, typename... Args>
 void from_json(const BasicJsonType& j, std::tuple<Args...>& t)
 {
     from_json_tuple_impl(j, t, index_sequence_for<Args...> {});
+}
+
+template <typename BasicJsonType, typename Key, typename Value,
+          typename = enable_if_t<not std::is_constructible<
+                                     typename BasicJsonType::string_t, Key>::value>>
+void from_json(const BasicJsonType& j, std::map<Key, Value>& m)
+{
+    if (JSON_UNLIKELY(not j.is_array()))
+    {
+        JSON_THROW(type_error::create(302, "type must be array, but is " + std::string(j.type_name())));
+    }
+    for (const auto& p : j)
+    {
+        if (JSON_UNLIKELY(not p.is_array()))
+        {
+            JSON_THROW(type_error::create(302, "type must be array, but is " + std::string(p.type_name())));
+        }
+        m.emplace(p.at(0).template get<Key>(), p.at(1).template get<Value>());
+    }
 }
 
 struct from_json_fn

--- a/single_include/nlohmann/json.hpp
+++ b/single_include/nlohmann/json.hpp
@@ -909,6 +909,7 @@ inline bool operator<(const value_t lhs, const value_t rhs) noexcept
 #include <ciso646> // and, not
 #include <forward_list> // forward_list
 #include <iterator> // inserter, front_inserter, end
+#include <map> // map
 #include <string> // string
 #include <tuple> // tuple, make_tuple
 #include <type_traits> // is_arithmetic, is_same, is_enum, underlying_type, is_convertible
@@ -1183,6 +1184,25 @@ template<typename BasicJsonType, typename... Args>
 void from_json(const BasicJsonType& j, std::tuple<Args...>& t)
 {
     from_json_tuple_impl(j, t, index_sequence_for<Args...> {});
+}
+
+template <typename BasicJsonType, typename Key, typename Value,
+          typename = enable_if_t<not std::is_constructible<
+                                     typename BasicJsonType::string_t, Key>::value>>
+void from_json(const BasicJsonType& j, std::map<Key, Value>& m)
+{
+    if (JSON_UNLIKELY(not j.is_array()))
+    {
+        JSON_THROW(type_error::create(302, "type must be array, but is " + std::string(j.type_name())));
+    }
+    for (const auto& p : j)
+    {
+        if (JSON_UNLIKELY(not p.is_array()))
+        {
+            JSON_THROW(type_error::create(302, "type must be array, but is " + std::string(p.type_name())));
+        }
+        m.emplace(p.at(0).template get<Key>(), p.at(1).template get<Value>());
+    }
 }
 
 struct from_json_fn

--- a/test/src/unit-conversions.cpp
+++ b/test/src/unit-conversions.cpp
@@ -1079,6 +1079,26 @@ TEST_CASE("value conversion")
                 j5.get<std::unordered_set<std::string>>();
             }
 
+            SECTION("std::map (array of pairs)")
+            {
+                std::map<int, int> m{{0, 1}, {1, 2}, {2, 3}};
+                json j6 = m;
+
+                auto m2 = j6.get<std::map<int, int>>();
+                CHECK(m == m2);
+
+                json j7 = {0, 1, 2, 3};
+                CHECK_THROWS_AS((j7.get<std::map<int, int>>()), json::type_error&);
+                CHECK_THROWS_WITH((j7.get<std::map<int, int>>()), "[json.exception.type_error.302] type must be array, but is number");
+
+                SECTION("superfluous entries")
+                {
+                  json j8 = {{0, 1, 2}, {1, 2, 3}, {2, 3, 4}};
+                  m2 = j8.get<std::map<int, int>>();
+                  CHECK(m == m2);
+                }
+            }
+
             SECTION("exception in case of a non-object type")
             {
                 CHECK_THROWS_AS((json().get<std::list<int>>()), json::type_error&);
@@ -1094,6 +1114,7 @@ TEST_CASE("value conversion")
                 CHECK_THROWS_WITH((json().get<std::vector<json>>()), "[json.exception.type_error.302] type must be array, but is null");
                 CHECK_THROWS_WITH((json().get<std::list<json>>()), "[json.exception.type_error.302] type must be array, but is null");
                 CHECK_THROWS_WITH((json().get<std::valarray<int>>()), "[json.exception.type_error.302] type must be array, but is null");
+                CHECK_THROWS_WITH((json().get<std::map<int, int>>()), "[json.exception.type_error.302] type must be array, but is null");
             }
         }
     }

